src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesClientProvider.java

```diff
@@ -0,0 +1,79 @@
+package org.csanchez.jenkins.plugins.kubernetes;
+
+import com.google.common.base.Objects;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.RemovalListener;
+
+import io.fabric8.kubernetes.client.KubernetesClient;
+
+import javax.annotation.CheckForNull;
+import java.io.IOException;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.CertificateEncodingException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Manages the Kubernetes client creation per cloud
+ */
+final class KubernetesClientProvider {
+
+    private static final Cache<String, Client> clients;
+
+    private static final Integer CACHE_SIZE;
+    private static final Integer CACHE_TTL;
+
+    static {
+        CACHE_SIZE = Integer.getInteger("org.csanchez.jenkins.plugins.kubernetes.clients.cacheSize", 10);
+        CACHE_TTL = Integer.getInteger("org.csanchez.jenkins.plugins.kubernetes.clients.cacheTtl", 60);
+        clients = CacheBuilder.newBuilder()
+                .maximumSize(CACHE_SIZE)
+                .expireAfterWrite(CACHE_TTL, TimeUnit.MINUTES)
+                .removalListener((RemovalListener<String, Client>) removalNotification -> {
+                    removalNotification.getValue().getClient().close();
```

Comment:
see https://google.github.io/guava/releases/23.0/api/docs/com/google/common/cache/RemovalNotification.html

> A notification of the removal of a single entry. The key and/or value may be null if they were already garbage collected.

so you need to check for null